workspaces/diff-engine/cli/src/main.rs

```diff
@@ -106,57 +109,105 @@ fn main() {
-          });
-        //dbg!("waiting for results");
-        let results = diff_comp
-          .await
-          .expect("diffing of interaction should be successful");
-        //dbg!("got results");
-
-        if let Some((results, tags)) = results {
-          for result in results {
-            //dbg!(&result);
-            if let Err(_) = results_sender
-              .send(ResultContainer::from((result, &tags)))
+    loop {
+      let next_action = tokio::select! {
+        // scheduling a diff task upon new input
+        next_interaction = interaction_lines.next() => {
+          if let None = next_interaction {
+            // end of input, stop looping
+            break;
+          }
+
+          let interaction_json_result = next_interaction.unwrap();
+
+          //dbg!("got next interaction");
+          let diff_permits = diff_scheduling_permits.clone();
+          let projection = spec_projection.clone();
+          let mut results_sender = results_sender.clone();
+          //dbg!("waiting for permit");
+          let diff_task_permit = diff_permits.acquire_owned().await;
+          //dbg!("got permit");
+
+          let diff_task = tokio::spawn(async move {
+            let diff_comp =
+              tokio::task::spawn_blocking::<_, Option<(Vec<InteractionDiffResult>, Tags)>>(move || {
+                let interaction_json =
+                  interaction_json_result.expect("can read interaction json line from stdin");
+                let TaggedInput(interaction, tags): TaggedInput<HttpInteraction> =
+                  match serde_json::from_str(&interaction_json) {
+                    Ok(tagged_interaction) => tagged_interaction,
+                    Err(parse_error) => {
+                      eprintln!("could not parse interaction json: {}", parse_error);
+                      return None;
+                    }
+                  };
+
+                Some((diff_interaction(&projection, interaction), tags))
+              });
+            //dbg!("waiting for results");
+            let results = diff_comp
               .await
-            {
-              panic!("could not write diff result to results channel");
-              // TODO: Find way to actually write error info
+              .expect("diffing of interaction should be successful");
+            //dbg!("got results");
+
+            if let Some((results, tags)) = results {
+              for result in results {
+                //dbg!(&result);
+                if let Err(_) = results_sender
+                  .send(ResultContainer::from((result, &tags)))
+                  .await
+                {
+                  panic!("could not write diff result to results channel");
+                  // TODO: Find way to actually write error info
+                }
+              }
             }
+
+            drop(diff_task_permit);
+          });
+
+          // dbg!("Scheduling task", diff_tasks.len() + 1);
+          diff_tasks.push(diff_task);
```

Comment:
This is a new essential part which allows the errors from the `diff_task` to be propagated. Before, we we'd just spawn the thread and let it run, with it's result never handled. Since these `select!`  handlers kind of work like a normal matching statement and are all handled on the same thread, we can write to `diff_tasks` while remaining accessible on the other branches as well.